### PR TITLE
Map panning fix

### PR DESCRIPTION
**Pull request by gegham-khachatryan**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25732

PROPOSAL: https://github.com/Expensify/App/issues/25732#issuecomment-1692767989

### Tests

1. Click the "+" icon next to the chat message text field. Click "Request Money" on the pop up menu that appears.
2. Check "Distance" tab on the Right Hand Panel that appears
3. Click the "Start" item from the list of waypoints.
4. In the text field that appears in the next screen, type "88 Kearny Street". Click the first option "88 Kearny Street, San Francisco, CA, USA" that appears.
5. Map camera flied to selected address where marker is visible.
6. Click Add Step button
7. Click on new added step 
8. In the text field that appears in the next screen, type "Golden Gate Park". Click the first option "Golden Gate Pkwy, Naples, FL, USA" that appears.
9. Verify that the 2 markers are on the map and camera moved to show both on the screen.
10. Added Finish step
11. Verify that the distance was drawn as expected. 

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Click the "+" icon next to the chat message text field. Click "Request Money" on the pop up menu that appears.
2. Check "Distance" tab on the Right Hand Panel that appears
3. Click the "Start" item from the list of waypoints.
4. In the text field that appears in the next screen, type "88 Kearny Street". Click the first option "88 Kearny Street, San Francisco, CA, USA" that appears.
5. Map camera flied to selected address where marker is visible.
6. Click Add Step button
7. Click on new added step 
8. In the text field that appears in the next screen, type "Golden Gate Park". Click the first option "Golden Gate Pkwy, Naples, FL, USA" that appears.
9. Verify that the 2 markers are on the map and camera moved to show both on the screen.
10. Added Finish step
11. Verify that the distance was drawn as expected. 

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/6061244/b437ad8f-ccc5-4d7a-b6a9-27a46534691e

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/6061244/a0d80461-063b-4c93-ae2e-d185d1becae4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/6061244/56d516ce-ce5f-4b1d-a9f6-0f19614f5e70

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/6061244/30f8006e-01b0-4d95-9490-9bf031db007b

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/6061244/8b4a8729-c441-4cbb-998a-6274b16ae651

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/6061244/4f4ab56d-f827-4714-b9e0-16b33dafc9ae

</details>

### CODE REVIEW

**Comment by gegham-khachatryan** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by gegham-khachatryan** [comment]
> Hi @gegham-khachatryan! Just a heads up. You're required to check all the checkboxes in the checklist. Also, you need to test and upload screen recordings of all of the platforms (not just iOS). Please sign the CLA as well!

@allroundexperts Testing done, screen recordings attached.

**Comment by hayata-suenaga** [comment]
is this one good to go [once this PR is merged](https://github.com/Expensify/react-native-x-maps/pull/32)?

**Comment by gegham-khachatryan** [comment]
> 

@neil-marcellini screen recordings here https://github.com/Expensify/App/pull/25977#issue-1867683017

**Comment by gegham-khachatryan** [comment]
> @gegham-khachatryan is this ready for review?

Yes @hayata-suenaga It's ready. I fixed also regression issue.

**Comment by JmillsExpensify** [comment]
How's testing going?

**Comment by gegham-khachatryan** [comment]
> Good so far!
> 
> @gegham-khachatryan Can you make sure that your commits are signed?
> 
> This is causing a check to fail.
@allroundexperts It seems not 😏 Is there any way to sign them?

**Comment by hayata-suenaga** [comment]
@gegham-khachatryan please check the bugs @allroundexperts found 🙇

**Comment by gegham-khachatryan** [comment]
> Did you notice the `Not Found Page` appearing for a split second?
> 
>  Screen.Recording.2023-08-31.at.3.03.09.AM.mov

It happens when transactions store updates complete before page transition end, And waypoint with current index is missing. `isFocused` state can fix this issue.

**Comment by gegham-khachatryan** [comment]
> Found a bug @gegham-khachatryan.
> 
>  Screen.Recording.2023-08-31.at.2.56.49.AM.mov

I found that `setCamera` or `fitBound` method is called properly but Mapbox SDK didn't reacted to this.

**Comment by gegham-khachatryan** [comment]
Also I found that Directions route not fetching and drawing. `shouldFetchRoute` is always false.  @thienlnam

**Comment by gegham-khachatryan** [comment]
@hayata-suenaga after syncing with main branch, the web UI broke (Left sidebar is missing). Direction Routes not loading
Screenshot 2023-08-31 at 11 35 11

**Comment by trjExpensify** [comment]
👋 @gegham-khachatryan, did you manage to run `npm install` okay to resolve that issue?

File: src/components/MapView/MapView.tsx
Comment (tgolen) [inline src/components/MapView/MapView.tsx:28]: Would you mind adding some comments around these effects so that it's clear what their purpose is and why they are there (same for below)?

File: src/components/MapView/MapView.tsx
Comment (tgolen) [inline src/components/MapView/MapView.tsx:73]: This variable doesn't look like it's necessary.

File: src/components/MapView/Direction.web.tsx
Comment (tgolen) [inline src/components/MapView/Direction.web.tsx:1]: Please add a comment to explain why there is a difference implementation for `web`

**Review by tgolen — COMMENTED** [review]
Looks like there is a bug when trying to drag the map around to the right (it's trying to change tabs).


https://github.com/Expensify/App/assets/1228807/ca9a2803-ec3a-4ac3-973f-c8d18f66a865

**Comment by hayata-suenaga** [comment]
@gegham-khachatryan please address PR reviews when you come back online 🙇

**Comment by thienlnam** [comment]
> Also I found that Directions route not fetching and drawing. shouldFetchRoute is always false. @thienlnam

I have this PR that makes the route fetching more reliable https://github.com/Expensify/App/pull/25980

**Comment by gegham-khachatryan** [comment]
@allroundexperts @trjExpensify Thanks `npm install` solve the issue.

**Comment by gegham-khachatryan** [comment]
> Probably, we should fix this console warning as well in this PR:
> 
> Screenshot 2023-08-31 at 3 00 00 AM

@allroundexperts I've added unique id to endpoint marker  https://github.com/Expensify/App/pull/25977/commits/2fbff6961b31ba2f9a335c4caa6c4160d1093962

**Comment by gegham-khachatryan** [comment]
> @gegham-khachatryan please address PR reviews when you come back online 🙇
> 
> Also, could you address this issue being addressed in [this x-maps PR](https://github.com/Expensify/react-native-x-maps/pull/35)? That library is, thanks to your effort, being closed soon. We might want to address the key issue described in the PR in this PR.

Fixed here https://github.com/Expensify/App/pull/25977/commits/2fbff6961b31ba2f9a335c4caa6c4160d1093962

File: src/components/MapView/Direction.web.tsx
Comment (gegham-khachatryan) [reply src/components/MapView/Direction.web.tsx]: Done

File: src/components/MapView/MapView.tsx
Comment (gegham-khachatryan) [reply src/components/MapView/MapView.tsx]: Updated

File: src/components/MapView/MapView.tsx
Comment (gegham-khachatryan) [reply src/components/MapView/MapView.tsx:32]: Updated

File: src/components/MapView/MapView.tsx
Comment (gegham-khachatryan) [reply src/components/MapView/MapView.tsx:78]: It's necessary, otherwise we having naming convention eslint error. I just copied all code from `react-native-x-maps` library. Thanks

File: src/components/MapView/MapView.web.tsx
Comment (gegham-khachatryan) [reply src/components/MapView/MapView.web.tsx:6]: Done

File: src/components/MapView/Direction.web.tsx
Comment (gegham-khachatryan) [reply src/components/MapView/Direction.web.tsx:6]: Done

**Comment by neil-marcellini** [comment]
> @gegham-khachatryan Let me know once this is fixed and I'll re-review this. I think the above mentioned issue is really an important one.

Since that's not strictly related to the bug let's fix it in a follow up. It's unlikely that people will make requests like this so we should tackle that bug with lower priority.

**Comment by gegham-khachatryan** [comment]
> > The following happens when places too far apart are chosen. I'm wondering if this is an issue in the mapbox sdk itself? 🤔
> > Screen.Recording.2023-08-31.at.3.04.21.AM.mov
> 
> @gegham-khachatryan Let me know once this is fixed and I'll re-review this. I think the above mentioned issue is really an important one.

I agree with you. Finally I could debug and fix issue ( I hope it's fixed ) @allroundexperts @neil-marcellini

**Comment by gegham-khachatryan** [comment]
@allroundexperts Now I can fall asleep 😂😂

**Comment by neil-marcellini** [comment]
Taking over review for Tim

**Comment by hayata-suenaga** [comment]
@gegham-khachatryan [for this bug](https://github.com/Expensify/App/pull/25977#pullrequestreview-1605156355), we already have [a separate issue](https://github.com/Expensify/App/issues/25422) created so you don't have to address it in this PR

**Comment by hayata-suenaga** [comment]
yes @allroundexperts are you not seeing them on both iOS and Web?

**Comment by hayata-suenaga** [comment]
@allroundexperts have you tried merging `main`?

**Comment by hayata-suenaga** [comment]
@allroundexperts can you test that on the latest main (without changes from this branch) to see if the issue is present?

**Review by allroundexperts — APPROVED** [review]
The panning behaviour works good on all devices. I encountered the following issues (which can be handled in a follow up PR):
1. The promise rejection warning when removing a waypoint. This happens only sometime.
2. The route marker is behaving in-consistently for me. Sometimes, it shows up while other times it does not.

And of course, the commits aren't signed. Since this is urgent, I guess we can let it go.

**Comment by hayata-suenaga** [comment]
> Something wrong with my setup I guess. Its the same with main!

I'm building test apps [here](https://github.com/Expensify/App/actions/runs/6043189178). Once finished, we can use test builds to test.

Also, I encountered a similar issue with iOS but I just needed to wait for a while for the route to be loaded

File: src/components/MapView/MapViewTypes.ts
Comment (neil-marcellini) [inline src/components/MapView/MapViewTypes.ts:6]: NAB: spaces above comments

File: src/components/MapView/MapView.tsx
Comment (neil-marcellini) [inline src/components/MapView/MapView.tsx:12]: NAB: Can we also specify here that the params are MapViewProps and the ref is a MapViewHandle, or is that already inferred?

Comment (neil-marcellini) [reply src/components/MapView/MapView.tsx:12]: Also I think we want this component defined with the function keyword.

File: src/components/MapView/Direction.tsx
Comment (neil-marcellini) [inline src/components/MapView/Direction.tsx:1]: NAB I prefer to call this Route

File: src/libs/actions/Transaction.js
Comment (neil-marcellini) [inline src/libs/actions/Transaction.js:96]: NAB:style: We like to always add brackets

**Review by neil-marcellini — DISMISSED** [review]
I left a bunch of non-blocking comments. Assuming it tests well I think it's good to go.

**Comment by gegham-khachatryan** [comment]
> Got this again:
> 
>  Screen.Recording.2023-09-01.at.2.54.04.AM.mov

Hi @allroundexperts It looks like an setup issue because I never seen that on my end.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by hayata-suenaga** [comment]
> When we select a very long route, no path line is shown. It might be fine.

yes it's fine 👍 not a blocker

> Add three points in US single state.
> Add last stop to somewhere in India.

No one drives from the US to India 😅 It's fine

**Comment by hayata-suenaga** [comment]
> But people do travel by air. Would that mean that they'll not be able to refund there ticket?

This distance request is mainly to reimburse expenses for gas when people travel by car

> Sounds good to me @hayata-suenaga. Also, I think we're increasing scope of this PR by keep on adding new things to it. Would be best to create a followup PR.

I agree 👍

**Comment by hayata-suenaga** [comment]
@gegham-khachatryan 

Screenshot 2023-09-01 at 7 33 27 AM

File: src/components/MapView/MapView.web.tsx
Comment (gegham-khachatryan) [reply src/components/MapView/MapView.web.tsx:21]: @parasharrajat I moved the code from [Expensify/react-native-x-maps](https://github.com/Expensify/react-native-x-maps). However, I noticed that they do not use useRef. Inside the Mapbox library component, they don't utilize forwardRef to provide access to higher-level components.

**Comment by gegham-khachatryan** [comment]
> @gegham-khachatryan
> 
> Screenshot 2023-09-01 at 7 33 27 AM

Fixed. I missed import during conflict resolution

**Comment by gegham-khachatryan** [comment]
@allroundexperts I cannot sign previous commits. I would greatly appreciate it if you could assist me in doing so.

**Comment by hayata-suenaga** [comment]
yes let's open another PR to implement @parasharrajat's suggestions

**Comment by hayata-suenaga** [comment]
@gegham-khachatryan ⬆️ 😃

**Comment by gegham-khachatryan** [comment]
@allroundexperts I looks impossible 😃 need to solve 463 conflicts, I don't know why)) I can create PR inside my fork, merge with squash, and open new PR with single commit to Expensify/main

**Review by neil-marcellini — APPROVED** [review]

**Comment by hayata-suenaga** [comment]
@gegham-khachatryan we have another set of ESLint issues now 😓

**Comment by gegham-khachatryan** [comment]
@hayata-suenaga https://github.com/Expensify/App/pull/26500
New PR opened. Lint errors fixed

**Comment by neil-marcellini** [comment]
Closed in favor of the new PR

File: src/pages/iou/WaypointEditor.js
Comment (gegham-khachatryan) [reply src/pages/iou/WaypointEditor.js:186]: @parasharrajat  There was an [issue](https://github.com/Expensify/App/pull/25977#issuecomment-1699909081) on waypoint delete. After deletion transactions updated before page closing animation is done, and during animation NotFound page become visible.

Comment (gegham-khachatryan) [reply src/pages/iou/WaypointEditor.js:186]: `Handle the issue on waypoint delete: When a waypoint is deleted, transactions are updated before the page closing animation is complete.  During the animation, the FullPageNotFoundView component  may become visible.` Something like this?

**Pull request closed without merging**
